tokio-net/src/driver/reactor/sharded_slab/page/mod.rs

```diff
@@ -0,0 +1,299 @@
+use super::{Pack, ScheduledIo, INITIAL_PAGE_SIZE, WIDTH};
+use crate::sync::{
+    atomic::{AtomicUsize, Ordering},
+    CausalCell,
+};
+
+pub(crate) mod slot;
+use self::slot::Slot;
+use std::fmt;
+
+/// A page address encodes the location of a slot within a shard (the page
+/// number and offset within that page) as a single linear value.
+#[repr(transparent)]
+#[derive(Copy, Clone, Eq, PartialEq, PartialOrd, Ord)]
+pub(crate) struct Addr {
+    addr: usize,
+}
+
+impl Addr {
+    const NULL: usize = Self::BITS + 1;
+    const INDEX_SHIFT: usize = INITIAL_PAGE_SIZE.trailing_zeros() as usize + 1;
+
+    pub(crate) fn index(self) -> usize {
+        // Since every page is twice as large as the previous page, and all page sizes
+        // are powers of two, we can determine the page index that contains a given
+        // address by shifting the address down by the smallest page size and
+        // looking at how many twos places necessary to represent that number,
+        // telling us what power of two page size it fits inside of. We can
+        // determine the number of twos places by counting the number of leading
+        // zeros (unused twos places) in the number's binary representation, and
+        // subtracting that count from the total number of bits in a word.
+        WIDTH - ((self.addr + INITIAL_PAGE_SIZE) >> Self::INDEX_SHIFT).leading_zeros() as usize
+    }
+
+    pub(crate) fn offset(self) -> usize {
+        self.addr
+    }
+}
+
+pub(super) fn size(n: usize) -> usize {
+    INITIAL_PAGE_SIZE * 2usize.pow(n as _)
+}
+
+impl Pack for Addr {
+    const LEN: usize = super::MAX_PAGES + Self::INDEX_SHIFT;
+
+    type Prev = ();
+
+    fn as_usize(&self) -> usize {
+        self.addr
+    }
+
+    fn from_usize(addr: usize) -> Self {
+        debug_assert!(addr <= Self::BITS);
+        Self { addr }
+    }
+}
+
+pub(in crate::driver) type Iter<'a> =
+    std::iter::FilterMap<std::slice::Iter<'a, Slot>, fn(&'a Slot) -> Option<&'a ScheduledIo>>;
+
+pub(crate) struct Local {
+    head: CausalCell<usize>,
+}
+
+pub(crate) struct Shared {
+    remote_head: AtomicUsize,
+    size: usize,
+    prev_sz: usize,
+    slab: CausalCell<Option<Box<[Slot]>>>,
+}
+
+impl Local {
+    pub(crate) fn new() -> Self {
+        Self {
+            head: CausalCell::new(0),
+        }
+    }
+
+    #[inline(always)]
+    fn head(&self) -> usize {
+        self.head.with(|head| unsafe { *head })
+    }
+
+    #[inline(always)]
+    fn set_head(&self, new_head: usize) {
+        self.head.with_mut(|head| unsafe {
+            *head = new_head;
+        })
+    }
+}
+
+impl Shared {
+    const NULL: usize = Addr::NULL;
+
+    pub(crate) fn new(size: usize, prev_sz: usize) -> Self {
+        Self {
+            prev_sz,
```

Comment:
Could you explain the reason `prev_sz` is stored? It seems fairly easy to compute (`size >> 1`). My gut is that it would be faster to compute than to store.

Reply:
@carllerche `prev_sz` is not just `size >> 1`, it's `(size >> 1) + (size >> 2) + (size >> 3) + ...` until we reach the initial page size. This isn't just the size of the page immediately before this one, it's the size of _all_ previous pages.

Reply:
Not a required change. It should be fairly straightforward to compute still. It sounds like: `size - (initial_page_size + 2)`